### ISSUE TITLE
Pick node by middle click

### ISSUE THREAD

**Issue by ClobberXD** [minetest/minetest#9781]
Continued from #2730, which has been locked (and loaded) as it packed a lot of heat.

****

The main reason I'm opening this issue again, is to discuss the implementation of this feature. From https://github.com/minetest/minetest/issues/2730#issuecomment-350390373,

> All items can optionally set a “pick item override” attribute in their definition. This can be a string or function. The function is only possible for nodes.
If it's a string, this is the itemstring the player will obtain.
If it's a function, this function will be called with arguments pos, node and must return the ItemStack to pick. By convention, the stack size should always be 1.
>
> This can be useful for internal items/blocks which the player must not obtain in the inventory. The idea is the same as for the not_in_creative_mode=1 group.

I really like Wuzzy's suggestion overall, but the `not_in_creative_inventory` part concerns me. It's very much possible for mods to add an extra field to their itemdef registrations, but what about backwards compat? For mods that don't have this field, how would this feature work?

The closest viable alternative I can think of is `drops` field - if this is defined (AFAIK it is, for most diggable `not_in_creative_inventory` nodes like wall/ceiling torches), this is the item to be returned. But this approach has its obvious flaws too - `drops` isn't always straight-forward, and can also be defined as a table instead, with attributes like drop chance. This field could potentially be omitted even for `not_in_creative_inventory` items (like admin stuff, maybe?).

Another option I can think of is to just omit the middle click behaviour for items that don't define Wuzzy's suggested field *and* set the group `not_in_creative_inventory`.

Opinions, suggestions, ideas, and inputs needed. I'm willing to have a go at this.

**Comment by LoneWolfHT** [comment]
If we add on_keypress, on_keyhold, etc to the MT API I feel like we could just add a 'third click type' and allow mods to define this behaviour themselves

**Comment by ClobberXD** [comment]
[Relevant IRC chatter](http://irc.minetest.net/minetest-dev/2020-04-30#i_5678991):

```
<ANAND> #9781
◄ShadowBot/#minetest-dev► https://github.com/minetest/minetest/issues/9781 -- Pick node by middle click
<+sfan5> ANAND: perhaps middle click should just have no functionality and be up to games to implement
<+Krock> ^
<+sfan5> it's not like it can really work client-side anyway since the client has to wait for the server to send the inventory change
<+rubenwardy> yeah
<+nerzhul_> middle click is used to split inventory stacks by 2 no ?
<+nerzhul_> or is this a game feature ?
<+Krock> nerzhul_: talking about walking-mode, not formspec-mode
<+sfan5> in Minecraft, if you are in creative mode it will automatically grab the node you are pointing at from the creative inventory and put it in the current slot
<+Krock> explains where the idea comes from
<ANAND> That makes sense
<ANAND> As a callback? on_pick?
<+Krock> some kind of user input callback
<+Krock> this flows kinda into the keypress callback requset
<ANAND> Yeah, that'd be nice
<+sfan5> ANAND: more like minetest.register_on_middleclicknode
<+sfan5> that is more in line with how existing node interaction works
<ANAND> Makes sense
<+nerzhul_> i think it's a good idea ot have a register for middleclick, as we have for other buttons
```

**Comment by paramat** [comment]
> All items can optionally set a “pick item override” attribute in their definition. This can be a string or function.

I do not like this hardcoding suggestion.
I also prefer just adding 'register on middle click' and leave it up to the game what to do with middle clicks.

**Comment by ClobberXD** [comment]
> I do not like this hardcoding suggestion.
> I also prefer just adding 'register on middle click' and leave it up to the game what to do with middle clicks.
> The 'pick node' feature belongs in games.

Yes, that's exactly what has been decided after the discussion on IRC (excerpt in previous comment).

**Comment by ClobberXD** [comment]
What would be the best way to inform the server of a middle-click event? Should this be sent as a part of `TOSERVER_INTERACT`?

Also, would it make sense to support rebinding the middleclick action to other keys? If yes, what should the action be called? Going by the previous discussion on IRC, this should be added as a callback, with no default implementation in the engine whatsoever, and allow mods to decide what they'd like to do with it. So where do we even start, when it comes to appropriately naming this action? We *could* stick with `keymap_middleclick`, but this obviously sounds confusing. :)

**Comment by sfan5** [comment]
> Should this be sent as a part of `TOSERVER_INTERACT`?

Yes

> Also, would it make sense to support rebinding the middleclick action to other keys?

You could leave that for your other PR that also implements LMB/RMB rebinding.

> So where do we even start, when it comes to appropriately naming this action?

There are two options here:
* Name the action `pick` so that users understand the purpose in the help and key menus. This obviously suggests a certain usage from the engine, not everyone might agree.
* Name it `middleclick` instead

**Comment by ClobberXD** [comment]
> You could leave that for your other PR that also implements LMB/RMB rebinding.

Okay, sounds good.

> Name the action pick so that users understand the purpose in the help and key menus. This obviously suggests a certain usage from the engine, not everyone might agree.

I like `pick` too, but I agree with the concern that the engine suggests a usage. Maybe we can overlook this concern, like we did in #7924 - where we settled upon the names "dig" and "place", in spite of the actions allowing for far more functionality than just digging and placing.

**Comment by johalun** [comment]
+1 for this. Coming from MC I really miss being able to middle click on any node and get the item in my hand. It's such a great quality of life improvement in both creative and survival.

**Comment by Haui1112** [comment]
+1 from me as well. Pickblock makes building a ton faster. Are there any updates or things we can help with?
